File: phases/03-data-structures/20-bloom-filters-cuckoo-filters-count-min-sketch/code/main.py
"""main.py — Bloom filter + Count-Min sketch in Python."""
from __future__ import annotations
import math


def mix64(x: int) -> int:
    x = (x + 0x9e3779b97f4a7c15) & 0xFFFFFFFFFFFFFFFF
    x = ((x ^ (x >> 30)) * 0xbf58476d1ce4e5b9) & 0xFFFFFFFFFFFFFFFF
    x = ((x ^ (x >> 27)) * 0x94d049bb133111eb) & 0xFFFFFFFFFFFFFFFF
    return x ^ (x >> 31)


class Bloom:
    def __init__(self, m: int, k: int) -> None:
        self.m, self.k = m, k
        self.bits = bytearray((m + 7) // 8)

    def _hashes(self, x: int) -> list[int]:
        return [mix64(x + i * 0xdeadbeef) % self.m for i in range(self.k)]

    def add(self, x: int) -> None:
        for h in self._hashes(x):
            self.bits[h // 8] |= 1 << (h & 7)

    def contains(self, x: int) -> bool:
        for h in self._hashes(x):
            if not (self.bits[h // 8] >> (h & 7)) & 1: return False
        return True


class CountMin:
    def __init__(self, w: int, d: int) -> None:
        self.w, self.d = w, d
        self.t = [[0] * w for _ in range(d)]

    def add(self, x: int, c: int = 1) -> None:
        for i in range(self.d):
            self.t[i][mix64(x + i * 0xcafef00d) % self.w] += c

    def estimate(self, x: int) -> int:
        return min(self.t[i][mix64(x + i * 0xcafef00d) % self.w] for i in range(self.d))


def main() -> None:
    b = Bloom(96000, 7)
    n_in, n_out = 10_000, 100_000
    for i in range(n_in): b.add(i)
    fp = sum(1 for i in range(n_in, n_in + n_out) if b.contains(i))
    fn = sum(1 for i in range(n_in) if not b.contains(i))
    theoretical = (1 - math.exp(-7 * n_in / 96000)) ** 7
    print(f"Bloom (m=96000, k=7, n={n_in}):")
    print(f"  false negatives: {fn} (must be 0)")
    print(f"  false positives: {fp} / {n_out} = {fp/n_out:.4f}  (theoretical {theoretical:.4f})")

    cm = CountMin(256, 4)
    for _ in range(1000): cm.add(42)
    for i in range(100):
        for _ in range(10): cm.add(i)
    print(f"\nCount-Min (w=256, d=4):")
    print(f"  estimate(42)  = {cm.estimate(42)}  (true 1010)")
    print(f"  estimate(7)   = {cm.estimate(7)}  (true 10)")
    print(f"  estimate(999) = {cm.estimate(999)}  (true 0)")


if __name__ == "__main__":
    main()

File: phases/03-data-structures/20-bloom-filters-cuckoo-filters-count-min-sketch/code/main.rs
//! main.rs — Bloom filter + Count-Min sketch in Rust.

fn mix64(mut x: u64) -> u64 {
    x = x.wrapping_add(0x9e3779b97f4a7c15);
    x = (x ^ (x >> 30)).wrapping_mul(0xbf58476d1ce4e5b9);
    x = (x ^ (x >> 27)).wrapping_mul(0x94d049bb133111eb);
    x ^ (x >> 31)
}

pub struct Bloom { bits: Vec<u8>, m: usize, k: usize }

impl Bloom {
    pub fn new(m: usize, k: usize) -> Self {
        Bloom { bits: vec![0u8; (m + 7) / 8], m, k }
    }
    fn h(&self, x: u64, i: usize) -> usize {
        (mix64(x.wrapping_add((i as u64).wrapping_mul(0xdeadbeef))) as usize) % self.m
    }
    pub fn add(&mut self, x: u64) {
        for i in 0..self.k {
            let h = self.h(x, i);
            self.bits[h / 8] |= 1 << (h & 7);
        }
    }
    pub fn contains(&self, x: u64) -> bool {
        (0..self.k).all(|i| {
            let h = self.h(x, i);
            (self.bits[h / 8] >> (h & 7)) & 1 == 1
        })
    }
}

pub struct CountMin { t: Vec<Vec<i64>>, w: usize, d: usize }

impl CountMin {
    pub fn new(w: usize, d: usize) -> Self {
        CountMin { t: vec![vec![0i64; w]; d], w, d }
    }
    fn h(&self, x: u64, i: usize) -> usize {
        (mix64(x.wrapping_add((i as u64).wrapping_mul(0xcafef00d))) as usize) % self.w
    }
    pub fn add(&mut self, x: u64, c: i64) {
        for i in 0..self.d {
            let idx = self.h(x, i);
            self.t[i][idx] += c;
        }
    }
    pub fn estimate(&self, x: u64) -> i64 {
        (0..self.d).map(|i| self.t[i][self.h(x, i)]).min().unwrap()
    }
}

fn main() {
    let mut b = Bloom::new(96000, 7);
    let n_in: u64 = 10_000;
    let n_out: u64 = 100_000;
    for i in 0..n_in { b.add(i); }
    let fp = (n_in..n_in + n_out).filter(|&x| b.contains(x)).count();
    let fn_ = (0..n_in).filter(|&x| !b.contains(x)).count();
    let theoretical = (1.0 - (-(7.0 * n_in as f64) / 96000.0).exp()).powi(7);
    println!("Bloom (m=96000, k=7, n={n_in}):");
    println!("  false negatives: {fn_}");
    println!("  false positives: {fp} / {n_out} = {:.4}  (theoretical {theoretical:.4})",
             fp as f64 / n_out as f64);

    let mut cm = CountMin::new(256, 4);
    for _ in 0..1000 { cm.add(42, 1); }
    for i in 0..100u64 { for _ in 0..10 { cm.add(i, 1); } }
    println!("\nCount-Min:");
    println!("  estimate(42)  = {}", cm.estimate(42));
    println!("  estimate(7)   = {}", cm.estimate(7));
    println!("  estimate(999) = {}", cm.estimate(999));
}
